File: packages/core/src/marginalia_core/ports/runtime.py
"""Runtime session supervision ports."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import UTC, datetime
from pathlib import Path
from typing import Protocol


def _utc_now() -> datetime:
    return datetime.now(UTC)


@dataclass(frozen=True, slots=True)
class RuntimeSessionRecord:
    """Persisted metadata for the active foreground runtime loop."""

    process_id: int
    session_id: str
    document_id: str
    command_language: str
    started_at: datetime = field(default_factory=_utc_now)
    entrypoint: str = "play"
    working_directory: Path | None = None


@dataclass(frozen=True, slots=True)
class RuntimeCleanupReport:
    """Best-effort cleanup result for a previously running foreground loop."""

    runtime_found: bool
    record_removed: bool
    terminated_process_ids: tuple[int, ...] = ()
    notes: tuple[str, ...] = ()

    @property
    def cleaned_up(self) -> bool:
        """Return whether startup removed or terminated prior runtime state."""

        return self.record_removed or bool(self.terminated_process_ids)


class RuntimeSupervisor(Protocol):
    """Manage the single supported foreground reading runtime."""

    def activate(self, record: RuntimeSessionRecord) -> None:
        """Persist the current runtime session record."""
        ...

    def current_runtime(self) -> RuntimeSessionRecord | None:
        """Return the currently registered runtime, if any."""
        ...

    def cleanup_existing_runtime(self, *, current_process_id: int) -> RuntimeCleanupReport:
        """Terminate and clear a previously registered runtime when safe."""
        ...

    def clear(self, *, process_id: int | None = None) -> None:
        """Remove the persisted runtime record."""
        ...

File: crates/marginalia-core/src/ports/runtime.rs
use chrono::{DateTime, Utc};
use std::path::PathBuf;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RuntimeSessionRecord {
    pub process_id: u32,
    pub session_id: String,
    pub document_id: String,
    pub command_language: String,
    pub started_at: DateTime<Utc>,
    pub entrypoint: String,
    pub working_directory: Option<PathBuf>,
    pub process_start_time: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RuntimeCleanupReport {
    pub runtime_found: bool,
    pub record_removed: bool,
    pub terminated_process_ids: Vec<u32>,
    pub notes: Vec<String>,
}

impl RuntimeCleanupReport {
    pub fn cleaned_up(&self) -> bool {
        self.record_removed || !self.terminated_process_ids.is_empty()
    }
}

pub trait RuntimeSupervisor {
    fn activate(&mut self, record: RuntimeSessionRecord);
    fn current_runtime(&self) -> Option<RuntimeSessionRecord>;
    fn cleanup_existing_runtime(&mut self, current_process_id: u32) -> RuntimeCleanupReport;
    fn clear(&mut self, process_id: Option<u32>);
}

#[cfg(test)]
mod tests {
    use super::RuntimeCleanupReport;

    #[test]
    fn cleaned_up_is_true_for_removed_record() {
        let report = RuntimeCleanupReport {
            runtime_found: true,
            record_removed: true,
            terminated_process_ids: Vec::new(),
            notes: Vec::new(),
        };

        assert!(report.cleaned_up());
    }
}
